File: ToyRobotSimulator/Models/Robot.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyRobotSimulator.UserControls;

namespace ToyRobotSimulator.Models;

public class Robot(int spaceW, int spaceH, RobotPreview? prv)
{
    public int posX = -1;
    public int posY = -1;
    public Direction facing = Direction.NORTH;

    public bool startMove = false;
    public string message = string.Empty;

    public event EventHandler<MessageEventArgs>? eh_Warning;

    public bool IsInWorkspace(int x, int y)
    {
        if (x >= 0 && x < spaceW && y >= 0 && y < spaceH)
        {
            prv?.Danger(false);
            return true;
        }

        eh_Warning?.Invoke(this, new MessageEventArgs("The robot falls out of the work space!"));
        prv?.Danger(true);
        return false;
    }

    public void Place(int x, int y, Direction dir)
    {
        // Make sure the robot is placed in the workplace
        if (!IsInWorkspace(x, y)) return;

        // Place the robot
        posX = x;
        posY = y;
        facing = dir;

        // Since the robot has been places, the following commands will be able to run
        startMove = true;

        prv?.Place(x, -y, facing);
    }

    public void Move()
    {
        if (!startMove) return;

        // Make sure the robot is placed in the workplace
        if (!IsInWorkspace(posX, posY)) return;

        // Record the current position
        int moveX = 0;
        int moveY = 0;

        // Move one unit to the facing facing
        // Origin(0,0) is the South West most corner
        switch (facing)
        {
            case Direction.NORTH:
                moveY = 1;
                break;
            case Direction.EAST:
                moveX = 1;
                break;
            case Direction.SOUTH:
                moveY = -1;
                break;
            case Direction.WEST:
                moveX = -1;
                break;
            default:
                break;
        }

        // If the robot will move out of the workspace, ignore this movement
        if (!IsInWorkspace(posX + moveX, posY + moveY)) return;

        // Otherwise, apply this movement
        posX += moveX;
        posY += moveY;

        prv?.Move(moveX, -moveY);
    }

    public void Rotate(string dir)
    {
        if (!startMove) return;

        // Make sure the robot is placed in the workplace
        if (!IsInWorkspace(posX, posY)) return;

        switch (dir)
        {
            case "LEFT":
                facing = (Direction)(((int)facing + 3) % 4);
                break;
            case "RIGHT":
                facing = (Direction)(((int)facing + 1) % 4);
                break;
            default:
                break;
        }

        prv?.Rotate(facing);
    }

    public void Report()
    {
        if (!startMove) return;

        message = $"{posX},{posY},{facing}";
    }
}

File: FastAPI + React-native + expo/Robot-Control-Simulator/backend/app/models/robot.py
from deps import Directions, Event
from typing import Literal

class robot:
    def __init__(self, workspace_width: int, workspace_height: int):
        self.workspace_width = workspace_width
        self.workspace_height = workspace_height
        self.pos_x = 0
        self.pos_y = 0
        self.dir = Directions.NORTH
        self.warning_event = Event()
        self.updated_event = Event()
        self.start_moving = False
        self.output = list[str]
        
    async def is_in_workspace(self,x: int, y: int) -> bool:
        if x >= 0 and x < self.workspace_width and y >= 0 and y < self.workspace_height:
            return True
        else:
            await self.warning_event.invoke("")
            return False
    
    async def place(self, x: str, y: str, facing: str):
        new_x = int(x)
        new_y = int(y)
        if not await self.is_in_workspace(new_x, new_y): return
        
        self.pos_x = new_x
        self.pos_y = new_y
        self.dir = Directions[facing]
        self.start_moving = True
        await self.updated_event.invoke("PLACE", {"x":self.pos_x, "y":self.pos_y, "f":self.dir.value})
        
    async def move(self):
        if not self.start_moving: return
        
        new_x = self.pos_x
        new_y = self.pos_y
        
        if self.dir == Directions.NORTH:
            new_y += 1
        elif self.dir == Directions.EAST:
            new_x += 1
        elif self.dir == Directions.SOUTH:
            new_y -= 1
        elif self.dir == Directions.WEST:
            new_x -= 1
        
        if await self.is_in_workspace(new_x, new_y):
            self.pos_x = new_x
            self.pos_y = new_y
            await self.updated_event.invoke("MOVE", {"x":self.pos_x, "y":self.pos_y})       
    
    async def rotate(self, to: Literal["LEFT", "RIGHT"]):
        if not self.start_moving: return
        
        if to == "LEFT":
            self.dir = Directions((self.dir.value - 1) % 4)
        elif to == "RIGHT":
            self.dir = Directions((self.dir.value + 1) % 4)
        await self.updated_event.invoke("ROTATE", self.dir.value)
            
    async def report(self):
        if not self.start_moving: return
        await self.updated_event.invoke("REPORT", f"{self.pos_x},{self.pos_y},{self.dir.name}")